src/java.base/share/classes/jdk/internal/foreign/layout/ValueLayouts.java

```diff
@@ -143,7 +142,16 @@ public int hashCode() {
         }
 
         @Override
-        abstract V dup(long bitAlignment, Optional<String> name);
+        final V dup(long bitAlignment, Optional<String> name) {
```

Comment:
question: instead of using a duplicator object, can't we have a new `dup` overload here with the required arguments, rewire the old `dup` to the new one, and then override as appropriate in all the subclasses? That would still allow you to define `withOrder` in a single place (which, AFAICS appears to be the main simplification?)

Reply:
> question: instead of using a duplicator object, can't we have a new `dup` overload here with the required arguments, rewire the old `dup` to the new one, and then override as appropriate in all the subclasses? That would still allow you to define `withOrder` in a single place (which, AFAICS appears to be the main simplification?)

This is exactly what I am proposing as an alternative in the PR header and perhaps that is easier to read. I will update the proposal accordingly.

Reply:
Sorry - I did miss that comment.